packages/alfa-url/src/url.ts

```diff
@@ -0,0 +1,364 @@
+   */
+  public get password(): Option<string> {
+    return this._password;
+  }
+
+  /**
+   * @see https://url.spec.whatwg.org/#concept-url-host
+   */
+  public get host(): Option<string> {
+    return this._host;
+  }
+
+  /**
+   * @see https://url.spec.whatwg.org/#concept-url-port
+   */
+  public get port(): Option<number> {
+    return this._port;
+  }
+
+  /**
+   * @see https://url.spec.whatwg.org/#concept-url-path
+   */
+  public get path(): Sequence<string> {
+    return this._path;
+  }
+
+  /**
+   * @see https://url.spec.whatwg.org/#concept-url-query
+   */
+  public get query(): Option<string> {
+    return this._query;
+  }
+
+  /**
+   * @see https://url.spec.whatwg.org/#concept-url-fragment
+   */
+  public get fragment(): Option<string> {
+    return this._fragment;
+  }
+
+  /**
+   * @see https://url.spec.whatwg.org/#include-credentials
+   */
+  public hasCredentials(): boolean {
+    return this._username.isSome() || this._password.isSome();
+  }
+
+  /**
+   * Remove the fragment portion of this URL.
+   *
+   * @remarks
+   * This method is useful for contexts in which the fragment portion of the URL,
+   * which isn't passed from client to server, is of no interest.
+   */
+  public withoutFragment(): URL {
+    if (this._fragment.isNone()) {
+      return this;
+    }
+
+    return new URL(
+      this._scheme,
+      this._username,
+      this._password,
+      this._host,
+      this._port,
+      this._path,
+      this._query,
+      None
+    );
+  }
+
+  /**
+   * @see https://url.spec.whatwg.org/#concept-url-equals
+   */
+  public equals(value: URL): boolean;
+
+  /**
+   * @see https://url.spec.whatwg.org/#concept-url-equals
+   */
+  public equals(value: unknown): value is this;
+
+  public equals(value: unknown): boolean {
+    return (
+      value instanceof URL &&
+      value._scheme === this._scheme &&
+      value._username.equals(this._username) &&
+      value._password.equals(this._password) &&
+      value._host.equals(this._host) &&
+      value._port.equals(this._port) &&
+      value._path.equals(this._path) &&
+      value._query.equals(this._query) &&
+      value._fragment.equals(this._fragment)
+    );
+  }
+
+  public hash(hash: Hash): void {
+    Hash.writeString(hash, this._scheme);
+    this._username.hash(hash);
+    this._password.hash(hash);
+    this._host.hash(hash);
+    this._port.hash(hash);
+    this._path.hash(hash);
+    this._query.hash(hash);
+    this._fragment.hash(hash);
+  }
+
+  public toJSON(): URL.JSON {
+    return {
+      scheme: this._scheme,
+      username: this._username.getOr(null),
+      password: this._password.getOr(null),
+      host: this._host.getOr(null),
+      port: this._port.getOr(null),
+      path: this._path.toArray(),
+      query: this._query.getOr(null),
+      fragment: this._fragment.getOr(null),
+    };
+  }
+
+  /**
+   * @see https://url.spec.whatwg.org/#concept-url-serializer
+   */
+  public toString(): string {
```

Comment:
This is a piece of code with a lot of logic and conditionals which is therefore super prone to regressions and would most benefit from unit tests.
I think this is OK as I don't think that bugs here are going to be tricky to track down, or likely to cause big trouble. So we can probably leave it as is for now and add tests if we encounter bugs later…

Reply:
Agreed 😄